massa-storage/src/endorsement_indexes.rs

```diff
@@ -8,89 +10,108 @@ use massa_models::{
-    pub(crate) fn batch_remove(&mut self, endorsement_ids: Vec<EndorsementId>) {
-        for id in endorsement_ids {
-            let endorsement = self
-                .endorsements
-                .remove(&id)
-                .expect("removing absent object from storage");
-            let creator = endorsement.creator_address;
-            let entry = self.index_by_creator.entry(creator).or_default();
-            entry.remove(&id);
-            if entry.is_empty() {
-                self.index_by_creator.remove(&creator);
-            }
+    /// - block_id: the block id to link the endorsements to
+    /// - endorsement_ids: the endorsements to link to the block
+    pub(crate) fn link_block(&mut self, block_id: &BlockId, endorsement_ids: &Vec<EndorsementId>) {
+        for endorsement_id in endorsement_ids {
+            // update to-block index
+            self.index_to_block
+                .entry(*endorsement_id)
+                .or_default()
+                .insert(*block_id);
         }
     }
 
-    /// Link a vector of endorsements to a block. Should be used in case of the block is added to the storage.
+    /// Unlinks a block from the endorsements. Should be used when a block is removed from storage.
     /// Arguments:
-    /// - block_id: the block id to link the endorsements to
-    /// - endorsement_ids: the endorsements to link to the block
-    pub(crate) fn link_endorsements_with_block(
+    /// - block_id: the block id to unlink the endorsements from
+    /// - endorsement_ids: the endorsements to unlink from the block
+    pub(crate) fn unlink_block(
         &mut self,
         block_id: &BlockId,
         endorsement_ids: &Vec<EndorsementId>,
     ) {
-        self.index_by_block
-            .entry(*block_id)
-            .or_default()
-            .extend(endorsement_ids);
+        for endorsement_id in endorsement_ids {
+            if let hash_map::Entry::Occupied(mut occ) = self.index_to_block.entry(*endorsement_id) {
+                occ.get_mut().remove(block_id);
+                if occ.get().is_empty() {
+                    occ.remove();
+                }
+            }
+        }
     }
 
-    /// Unlink a vector of endorsements from a block. Should be used in case of the block is removed from the storage.
+    /// Remove a endorsement, remove from the indexes and made some clean-up in indexes if necessary.
     /// Arguments:
-    /// - block_id: the block id to unlink the endorsements from
-    pub(crate) fn unlink_endorsements_from_block(&mut self, block_id: &BlockId) {
-        self.index_by_block.remove(block_id);
+    /// - endorsement_id: the endorsement id to remove
+    pub(crate) fn remove(&mut self, endorsement_id: &EndorsementId) -> Option<WrappedEndorsement> {
+        // update to-block index
+        self.index_to_block.remove(endorsement_id);
+
+        if let Some(e) = self.endorsements.remove(endorsement_id) {
+            // update creator index
+            self.index_by_creator
```

Comment:
Maybe delete index by creator entry if empty at the end (applies to all the others (block, operation))

Reply:
done